File: python/pia/resources/macro.py
"""Official PIA SDK - Macro & Market Sentiment Resource."""

from __future__ import annotations

import urllib.parse
from typing import Any, Optional

from ..errors import ValidationError
from ..transport import AsyncTransport, SyncTransport
from ..types import (
    CentralBankStanceResponse,
    CotReportResponse,
    FearGreedData,
    FearGreedHistoryResponse,
)


class MacroResource:
    """Synchronous Macro, Sentiment & Positioning resource."""

    def __init__(self, transport: SyncTransport) -> None:
        self._transport = transport

    def get_fear_greed(self) -> FearGreedData:
        """Retrieves current Fear & Greed Index score and sentiment rating."""
        data = self._transport.request("/api/v1/fear-greed", method="GET")
        return FearGreedData.from_dict(data)

    def get_fear_greed_history(self) -> FearGreedHistoryResponse:
        """Retrieves historical Fear & Greed Index time series."""
        data = self._transport.request("/api/v1/fear-greed/history", method="GET")
        return FearGreedHistoryResponse.from_dict(data)

    def get_cot(self, symbol: str) -> CotReportResponse:
        """Fetches CFTC Commitment of Traders (COT) positioning report for a symbol."""
        if not symbol or not symbol.strip():
            raise ValidationError("Symbol must be a non-empty string.", param_name="symbol")
        clean = symbol.strip().upper()
        encoded = urllib.parse.quote(clean, safe="")
        data = self._transport.request(f"/api/v1/cot/symbol/{encoded}", method="GET")
        return CotReportResponse.from_dict(data)

    def get_central_bank_stance(self, bank: str) -> CentralBankStanceResponse:
        """Fetches policy stance and interest rate assessment for a central bank."""
        if not bank or not bank.strip():
            raise ValidationError("Bank must be a non-empty string.", param_name="bank")
        clean = bank.strip().lower()
        encoded = urllib.parse.quote(clean, safe="")
        data = self._transport.request(f"/api/v1/central-banks/{encoded}/stance", method="GET")
        return CentralBankStanceResponse.from_dict(data)


class AsyncMacroResource:
    """Asynchronous Macro, Sentiment & Positioning resource."""

    def __init__(self, transport: AsyncTransport) -> None:
        self._transport = transport

    async def get_fear_greed(self) -> FearGreedData:
        """Asynchronously retrieves current Fear & Greed Index score."""
        data = await self._transport.request("/api/v1/fear-greed", method="GET")
        return FearGreedData.from_dict(data)

    async def get_fear_greed_history(self) -> FearGreedHistoryResponse:
        """Asynchronously retrieves historical Fear & Greed Index time series."""
        data = await self._transport.request("/api/v1/fear-greed/history", method="GET")
        return FearGreedHistoryResponse.from_dict(data)

    async def get_cot(self, symbol: str) -> CotReportResponse:
        """Asynchronously fetches CFTC Commitment of Traders (COT) positioning."""
        if not symbol or not symbol.strip():
            raise ValidationError("Symbol must be a non-empty string.", param_name="symbol")
        clean = symbol.strip().upper()
        encoded = urllib.parse.quote(clean, safe="")
        data = await self._transport.request(f"/api/v1/cot/symbol/{encoded}", method="GET")
        return CotReportResponse.from_dict(data)

    async def get_central_bank_stance(self, bank: str) -> CentralBankStanceResponse:
        """Asynchronously fetches monetary policy stance for a central bank."""
        if not bank or not bank.strip():
            raise ValidationError("Bank must be a non-empty string.", param_name="bank")
        clean = bank.strip().lower()
        encoded = urllib.parse.quote(clean, safe="")
        data = await self._transport.request(f"/api/v1/central-banks/{encoded}/stance", method="GET")
        return CentralBankStanceResponse.from_dict(data)

File: typescript/src/resources/macro.ts
/**
 * Official PIA SDK - Macro & Market Sentiment Resource
 */

import { ValidationError } from "../errors";
import type { HttpTransport } from "../http/transport";
import type {
  CentralBankStanceResponse,
  CotReportResponse,
  FearGreedData,
  FearGreedHistoryResponse,
  RequestOptions,
} from "../types";

export class MacroResource {
  constructor(private readonly transport: HttpTransport) {}

  /**
   * Retrieves the current Fear & Greed Index score and sentiment rating.
   */
  public async getFearGreed(options?: RequestOptions): Promise<FearGreedData> {
    return this.transport.request<FearGreedData>(
      "/api/v1/fear-greed",
      "GET",
      undefined,
      options
    );
  }

  /**
   * Retrieves historical Fear & Greed Index time series.
   */
  public async getFearGreedHistory(
    options?: RequestOptions
  ): Promise<FearGreedHistoryResponse> {
    return this.transport.request<FearGreedHistoryResponse>(
      "/api/v1/fear-greed/history",
      "GET",
      undefined,
      options
    );
  }

  /**
   * Fetches CFTC Commitment of Traders (COT) institutional positioning report for a symbol.
   *
   * @param symbol Commodity/Currency/Index symbol (e.g. "GOLD", "WTI", "EURUSD", "SPX")
   */
  public async getCot(
    symbol: string,
    options?: RequestOptions
  ): Promise<CotReportResponse> {
    if (!symbol || typeof symbol !== "string" || symbol.trim() === "") {
      throw new ValidationError("Symbol must be a non-empty string.", "symbol");
    }

    const clean = symbol.trim().toUpperCase();
    const raw = await this.transport.request<any>(
      `/api/v1/cot/symbol/${encodeURIComponent(clean)}`,
      "GET",
      undefined,
      options
    );
    const reports = raw?.reports || raw?.data || raw?.items || [];
    return { ...raw, symbol: clean, reports: Array.isArray(reports) ? reports : [] };
  }

  /**
   * Fetches monetary policy stance and interest rate assessment for a central bank.
   *
   * @param bank Bank code (e.g. "fed", "ecb", "boj", "bi", "boe")
   */
  public async getCentralBankStance(
    bank: string,
    options?: RequestOptions
  ): Promise<CentralBankStanceResponse> {
    if (!bank || typeof bank !== "string" || bank.trim() === "") {
      throw new ValidationError("Bank must be a non-empty string.", "bank");
    }

    const clean = bank.trim().toLowerCase();
    const raw = await this.transport.request<any>(
      `/api/v1/central-banks/${encodeURIComponent(clean)}/stance`,
      "GET",
      undefined,
      options
    );
    return raw?.data && typeof raw.data === "object" ? { ...raw.data, bank: raw.data.bank || clean } : raw;
  }
}
